### ISSUE TITLE
3+D Arrays (nested structs)

### ISSUE THREAD

**Issue by pbower** [pola-rs/polars#12391]
## Feature Request

**Is your feature request related to a problem? Please describe.**
Currently, Polars does not support a struct-like data type for organising and manipulating sequences in time series data, and 'structs in structs'. This limitation makes it challenging to perform sequence-level computations and analysis efficiently, especially for multivariate time series datasets, and potentially multi-dimensional arrays.

**Describe the solution you'd like**
I propose adding support for a new data type, or enhancing existing types, to allow each sequence in a time series dataset to be treated as a single struct-like entity. This feature would enable easier and more efficient computations across each sequence, simplifying tasks such as applying functions to each sequence, computing aggregate statistics, and performing complex reshaping and indexing operations.
Additionally, a 'struct in a struct' concept could be considered. so that one can simply put a struct in a struct, and then still be able to do respective computations on the whole. 
This would then mean that a typical dataframe of say a customer, can store complex objects within a single cell (entry) for that customer, and complex respective computations can be carried out between those customers.

**Expected benefits include:**
- Better organisation and intuitive structure for time series datasets.
- Simplified and more efficient computation on a per-sequence basis.
- Enhanced support for advanced time series analysis techniques, such as window functions and lagged features.
- Improved compatibility with time series models in machine learning and deep learning.
- Highly compact, intuitive but semantic storage of complex data structures within a typical table like structure.
- A further edge on other data processing libraries.

**Considerations:**
- Performance impact, particularly regarding memory usage and computational efficiency.
- Complexity in implementation and data manipulation, requiring sophisticated functions and methods.

**Describe alternatives you've considered**
An alternative could be using existing data structures (like multi-dimensional arrays) and performing manual transformations. However, this approach lacks the intuitiveness and efficiency that a struct-like entity could provide.
Additionally, the ability to parallelise these computations offers large benefits, which, although can benefit from GPU-level computations, may still be effective here, although no GPU is something that may be an issue.

**Additional context**

Example:

# 10 sequences (t), each length 50 with 5 features
array([[[0.74375059, 0.19992437, 0.60312562, 0.94348685, 0.36559211],
        [0.96754122, 0.3808764 , 0.7105061 , 0.61281031, 0.34451137],
        [0.50689357, 0.55766007, 0.21734407, 0.71649191, 0.46734172],
        ...,
        [0.64570008, 0.89202895, 0.54614398, 0.19419067, 0.86965118],
        [0.28621092, 0.93728454, 0.69478809, 0.03784738, 0.79841516],
        [0.89863023, 0.33521001, 0.44677525, 0.64313013, 0.01733279]],

       [[0.55428501, 0.90845921, 0.9227793 , 0.51872951, 0.74231726],
        [0.35059427, 0.58296821, 0.97312522, 0.10604116, 0.69071185],
        [0.08723781, 0.47445343, 0.49601322, 0.63265866, 0.22051548],
        ...,
        [0.18570326, 0.76094031, 0.69872599, 0.22681627, 0.53615415],
        [0.49512315, 0.22309264, 0.48170262, 0.44178848, 0.49905379],
        [0.24305168, 0.15965623, 0.45301117, 0.37682866, 0.50520798]],
.....

Each 'block' becomes 1 Column in the struct, and potentially utility functions to 'explode' into separate columns within the dataframe for convenience.

Thank you for considering this feature request.

Please Note: I am open to to working on this feature in future, but it will be 6+ months or so until I am at a high-proficiency level with Rust as I only recently started using the language.

**Comment by rancomp** [comment]
I find this issue interesting because I believe discussing it would help everyone better understand Polars' philosophy.

Two questions, one specific for you and one general for the experts here:

1. Regarding time-series data, is there anything specific you want to achieve but find the current API too restrictive? I work with time-series data myself and find Polars a breeze.
2. I understand that if the data is hierarchical or sparse, there may be specialized ways to store it. Could these be implemented within the Polars backend, or might they be considered too specialized?

Aside from these questions, which would help us address practical difficulties, there's still the question of the Polars way of processing data.
In my view, data can generally be represented as a matrix, where the rows are different data or records, and the columns are the features of each datum/record. This matrix may be sparse and contain many repeating values.

Data types such as `Array, List, Struct` are essentially adding one dimension to an already high-dimensional data. In other words, the data can be represented by exploding the `Array` and adding another column with labels for each group. This doesn't introduce much of a difference – an 'Array-within-an-Array'? That's just two more columns.

I'm a fan of Polars' `Array, List, and Struct` entities. I'm using them
- Mostly for aggregating data into one column
- Occasionally for performing calculations
- Never for retrieving values

However, for calculations, operations like `group_by, over, and rolling` are by and large the workhorses for aggregate calculations. To achieve a 'struct-within-a-struct,' one would simply need to `group_by` two columns. Yes, the `group-by` object would be hidden from plain sight, but if you really want, you could iterate over and manipulate each group. Consider the alternative where each cell of the table has a matrix in it. That sounds chaotic.

Personally, aside from very specialized cases, I find it awkward to work with a multi-index because more often than not, I simply unstack the whole thing and work with a large table. Some libraries allow representing hierarchical or longitudinal data more directly, e.g., pandas' multi-indexed DataFrame. This multi-indexed representation is, in my opinion, only superficially better. It's nice when you want to make the hierarchical nature explicit and useful when you want to retrieve data quickly.

Note, [the coming-from-pandas section in the user guide ](https://pola-rs.github.io/polars/user-guide/migration/pandas/) opens up with `Polars does not have a multi-index/index`.

**Comment by pbower** [comment]
Hi Rancomp,

Thank you for your thoughtful response. It helps to clarify Polars' philosophy and its practical implications.

1. **Time Series Data**: Indeed, Polars offers robust features for time series analysis. My initial thought was exploring if struct-like entities could offer additional flexibility or compactness, but I appreciate the current capabilities Polars provides in this domain.

2. **Struct within a Struct - Compactness, Semantics, and Calculation**:
   - **Compactness**: The idea of nested structs (or a struct within a struct) would allow encapsulating complex, multi-dimensional data in a more compact form. This is particularly beneficial when dealing with data that is naturally hierarchical or nested, such as documents with nested sections, or entities with sub-entities (like companies and their departments), enabling the pulling of disparate relational tables into a single organising construct.
   - **Semantics**: Nested structures can provide a semantically richer representation of data. For example, consider a dataset of patients where each patient has multiple medical records. Representing each patient's records as a nested struct within a patient struct can mirror the real-world relationship more intuitively.
   - **Calculation and Data Integrity**: Nested structs can ensure data integrity and context preservation, especially when performing transformations or calculations. Operations on a nested struct can be inherently scoped to that structure, reducing the risk of inadvertently mixing or misaligning data across different entities.

   For instance, storing an entire image or a video as a multidimensional object within a single cell, is a compelling use case. It maintains the integrity of the data object (like an image) while allowing for operations on the whole object without the need to explode it into multiple rows or columns. This can be particularly advantageous in scenarios like image processing (and pre-processing) within a larger dataset, where each row represents a different entity (like a product) with an associated image.

In conclusion, while I understand and appreciate the efficiency and simplicity of Polars' current data representation, considering nested structures could open up new possibilities for handling complex data types more naturally and intuitively, especially in scenarios where data integrity and semantic representation are crucial.

Thank you again for considering these perspectives.

**Comment by mkleinbort-ic** [comment]
If I understand the issue I think it'd be mainly "resolved" by the following feature (which I would like):

> struct columns should implement a `.unnest()` method. eg. `pl.Expr.struct.unnest()` that expand the fields into new columns

Here is an example of how that could be used:

```python
# Take a struct of values, rank all the values, and return them to a struct

df = pl.DataFrame({
    'x': [
        {'A':1, 'B':2, 'C':3},
        {'A':2, 'B':1, 'C':2},
        {'A':3, 'B':0, 'C':5},
          ]
}).with_columns(
  x_ranked = pl.col('x').unnest().rank().pipe(pl.struct)
)
```

Might raise a feature request for this.


Also, note that the expression ` x_ranked = pl.col('x').unnest().rank().pipe(pl.struct)` is actually pretty hard to replicate in the current Polars API - best I could come up with was this DataFrame level method:

```python
(df
    .pipe(lambda self: pl.concat([
            self, 
            self.select('x').unnest('x').select(pl.all().rank().pipe(pl.struct).alias('x_ranked'))],
      how='horizontal')
)
```

**Comment by cmdlineluser** [comment]
@mkleinbort-ic You can do it at the "Expr" level if it helps:

```python
df.with_columns(x_ranked = 
   pl.col("x").map_batches(lambda x:
      x.struct.unnest().with_columns(pl.all().rank())
       .to_struct("")
   )
)
```

This is essentially how `.map_dict` is (currently) implemented.

https://github.com/pola-rs/polars/blob/de2a5efba375a2a47d79a1be1c77213778fc287c/py-polars/polars/expr/expr.py#L9332

**Comment by mitstake** [comment]
I just started a discussion on how polars could support multidimensional array style operations, that might be of interest and address the issues you're describing here:

https://github.com/pola-rs/polars/issues/23938
